lib/libc/minimal/source/stdlib/qsort.c

```diff
@@ -59,28 +73,34 @@ static void sift_down(void *base, int start, int end, size_t size, comp3_t comp,
 	}
 }
 
-static void heapify(void *base, int nmemb, size_t size, comp3_t comp, void *comp_arg)
+static void heapify(void *base, int nmemb, size_t size, void *comp, void *comp_arg,
+		    bool comp_fn_has_3)
 {
 	int start;
 
 	for (start = parent(nmemb - 1); start >= 0; --start) {
-		sift_down(base, start, nmemb, size, comp, comp_arg);
+		sift_down(base, start, nmemb, size, comp, comp_arg, comp_fn_has_3);
 	}
 }
 
-static void heap_sort(void *base, int nmemb, size_t size, comp3_t comp, void *comp_arg)
+static void heap_sort(void *base, int nmemb, size_t size, void *comp, void *comp_arg,
+		      bool comp_fn_has_3)
 {
 	int end;
 
-	heapify(base, nmemb, size, comp, comp_arg);
+	heapify(base, nmemb, size, comp, comp_arg, comp_fn_has_3);
 
 	for (end = nmemb - 1; end > 0; --end) {
 		byteswp(A(end), A(0), size);
-		sift_down(base, 0, end, size, comp, comp_arg);
+		sift_down(base, 0, end, size, comp, comp_arg, comp_fn_has_3);
 	}
 }
+void qsort_r(void *base, size_t nmemb, size_t size, compar_d_fn_t compar, void *arg)
```

Comment:
I'm ok with removing the static inline version of qsort from the minimal libc `stdlib.h`. No need to add any typedefs anywhere really, even in `qsort.c`.

The qsort implementation can be simplified by instantiating a `struct qsort_comp` in the bodies of `qsort()` and `qsort_r()`. With this, the `comp_arg` parameter can be removed from `heap_sort()` and `sift_down()`.

```
struct qsort_comp {
	bool has3;
	void *arg;
	union {
		int (*comp2)(const void *, const void *);
		int (*comp3)(const void *, const void *, void *);
	};
};

static inline int comp(struct qsort_comp *cmp, const void *a, const void *b)
{
	if (cmp->has3) {
		return cmp->comp3(a, b, cmp->arg);
	}

	return cmp->comp2(a, b);
}
```

Reply:
@cfriedt Thanks for the suggestion, and I have now reworked the PR as per your suggestion. I hope it is acceptable now, but let me know what you think.